### ISSUE TITLE
Extend `shield()` to support more complex types

### ISSUE THREAD

**Issue by pencilcheck** [telefunc/telefunc#97]
I have been using drizzle orm for database typing generation. I have a enum I want to expose as a function parameter. I have created the type but I'm not sure if I can pass it onto the field, as it seems like shield is expecting a certain type of typing input in order for it to work.

I know how to specify enum from shield, but is it possible to allow custom typing such as drizzle to be used instead?

Example
```
import { recordType } from '../../../server/drizzle/schema';

type RecordType = typeof recordType

const t = shield.type;

const hello = shield(
  [{
    recordType: t.or(
      t.const('HIDDEN'),
      t.const('PENDING'),
      t.const('CONFIRMED')
    ), <<<<---- replace it with RecordType?
    year: t.number,
    timezone: t.string
  }],
```

**Comment by brillout** [comment]
I agree it would be nice, but it's far from trivial, and not even possible if drizzle doesn't support type inference.

As for adding support for Zod, it's a consideration but so far `shield()` seems to be working for the (vast?) majority of use cases. Especially when following the recommendation depicted at https://telefunc.com/event-based.

Closing as I don't think it should be a priority right now: so far, I don't see the added value to be worth the effort. Feel free to object and elaborate.

Contribution or sponsoring welcome.

**Comment by pencilcheck** [comment]
> I agree it would be nice, but it's far from trivial, and not even possible if drizzle doesn't support type inference.
> 
> As for adding support for Zod, it's a consideration but so far `shield()` seems to be working for the (vast?) majority of use cases. Especially when following the recommendation depicted at https://telefunc.com/event-based.
> 
> Closing as I don't think it should be a priority right now: so far, I don't see the added value to be worth the effort. Feel free to object and elaborate.
> 
> Contribution or sponsoring welcome.

I don't know what is the right solution, I am just thinking if there is a better integration of type validation with other libraries such as drizzle or zod.

**Comment by brillout** [comment]
> better integration [...] with other libraries such as drizzle

I agree that'd be nice. But, again, I don't know if it's possible: does Drizzle types support runtime validation like Zod does?

As for Zod, I still don't see much of an added value compared to `shield()`. But feel free to elaborate if you think otherwise. (I understand Zod has semantic validation such as email validation, but such semantic validation doesn't add any value regarding TypeScript runtime validation.)

**Comment by brillout** [comment]
I'm not against this at all. I'm just waiting for a good reason for implementing this. Labeling this as `low-prio` until then.

**Comment by brillout** [comment]
Actually, form data is a valid use case for this as explained at https://github.com/brillout/telefunc/issues/105.

Removing `low-prio 🐌`, but there are still other higher priorities: https://github.com/brillout/telefunc/issues?q=is%3Aissue+is%3Aopen+label%3A%22high-prio+%F0%9F%92%AB%22.

That said, contribution or sponsoring welcome.

If someone wants to contribute, we can start by having a look at how tRPC does it.

**Comment by brillout** [comment]
I'm thinking, maybe an alternative could be to allow users to use Telefunc's `shield()` independently of Telefunc. There are zillions of runtime validators, can't we just implement one that works for the vast majority of use cases? Or are there really that many use cases that justify being able to pick a validator that fits best? (As for semantic validation, the user could use a semantic validator in addition to using `shield()`.)

**Comment by pencilcheck** [comment]
shield handles basic use cases well, but it doesn't seem to allow defining more complex use cases such as array structure, nested object structure, etc. If shield can be extended with other custom validator and it works well then it would be very useful.

But the use case this post is about isn't about extensibility. I'm asking if it is possible if I can have my typescript types to be used with shield.

**Comment by brillout** [comment]
@pencilcheck Can you elaborate your use case for needing complex Telefunction arguments? Are you following https://telefunc.com/event-based?

**Comment by brillout** [comment]
I did some triage about this topic, the feature request now lives at:
- https://github.com/telefunc/telefunc/issues/244

As for the original issue (extending `shield()` to support more complex types), I consider it a bug: Telefunc should support any type that TypeScript supports. If not then we should fix it: reproduction welcome, I'll re-open this issue then.

> I know how to specify enum from shield, but is it possible to allow custom typing such as drizzle to be used instead?

How about [this pattern](https://github.com/telefunc/telefunc/issues/105#issuecomment-4081408482)?
